auth_pysaml/controllers/main.py

```diff
@@ -0,0 +1,234 @@
+                var r = '/' + l.search;
+                if(q.length !== 0) {
+                    var s = l.search ? (l.search === '?' ? '' : '&') : '?';
+                    r = l.pathname + l.search + s + q;
+                }
+                window.location = r;
+            </script></head><body></body></html>"""
+        return func(self, req, **kw)
+
+    return wrapper
+
+
+# ----------------------------------------------------------
+# Controller
+# ----------------------------------------------------------
+
+
+class SAMLLogin(Home):
+    def _list_saml_providers_domain(self):
+        return []
+
+    def list_providers(self):
+        providers = (
+            request.env["auth.saml.provider"]
+            .sudo()
+            .search_read(self._list_saml_providers_domain())
+        )
+
+        for provider in providers:
+            # Compatibility with auth_oauth/controllers/main.py in order to
+            # avoid KeyError rendering template_auth_oauth_providers
+            provider.setdefault("auth_link", "")
+        return providers
+
+    @http.route()
+    def web_login(self, *args, **kw):
+        ensure_db()
+        if (
+            request.httprequest.method == "GET"
+            and request.session.uid
+            and request.params.get("redirect")
+        ):
+
+            # Redirect if already logged in and redirect param is present
+            return http.redirect_with_hash(request.params.get("redirect"))
+
+        providers = self.list_providers()
+
+        response = super().web_login(*args, **kw)
+        if response.is_qweb:
+            error = request.params.get("saml_error")
+            if error == "no-signup":
+                error = _("Sign up is not allowed on this database.")
+            elif error == "access-denied":
+                error = _("Access Denied")
+            elif error == "expired":
+                error = _(
+                    "You do not have access to this database. Please contact"
+                    " support."
+                )
+            else:
+                error = None
+
+            response.qcontext["providers"] = providers
+
+            if error:
+                response.qcontext["error"] = error
+
+        return response
+
+
+class AuthSAMLController(http.Controller):
+    def _get_saml_extra_relaystate(self):
+        """
+        Compute any additional extra state to be sent to the IDP so it can
+        forward it back to us. This is called RelayState.
+
+        The provider will automatically set things like the dbname, provider
+        id, etc.
+        """
+
+        redirect = request.params.get("redirect") or "web"
+        if not redirect.startswith(("//", "http://", "https://")):
+            redirect = "{}{}".format(
+                request.httprequest.url_root,
+                redirect[1:] if redirect[0] == "/" else redirect,
+            )
+
+        state = {
+            "r": werkzeug.url_quote_plus(redirect),
```

Comment:
In order to avoid the following warning:

``` 
'werkzeug.url_quote_plus' is deprecated and will be removed in
Werkzeug 1.0. Use 'from werkzeug.urls import url_quote_plus
```

you should import the metod as follows:

`from werkzeug.urls import url_quote_plus`

And then use it:

```suggestion
            "r": url_quote_plus(redirect),
```